File: backend/utils/severity_calculator.py
TYPE_SEVERITY_WEIGHTS = {
    "Scratch": 30.0,
    "Surface Defect": 40.0,
    "Dent": 60.0,
    "Crack": 90.0,
    "Cracked Screen": 95.0,
    "Missing Component": 95.0
}


def calculate_severity(
    defect_type: str,
    size_mm2: float,
    is_functional_location: bool,
    confidence: float
) -> dict:
    size_score = min(100.0, (size_mm2 / 100.0) * 100.0)
    location_score = 90.0 if is_functional_location else 35.0
    type_score = TYPE_SEVERITY_WEIGHTS.get(defect_type, 50.0)
    confidence_score = confidence * 100.0

    overall_score = round(
        (size_score * 0.30) +
        (location_score * 0.25) +
        (type_score * 0.25) +
        (confidence_score * 0.20),
        1
    )

    if overall_score >= 80.0:
        level = "Critical"
        recommendation = "Reject product and trigger immediate quality inspection workflow."
    elif overall_score >= 60.0:
        level = "High"
        recommendation = "Significant quality issue. Rework or repair recommended."
    elif overall_score >= 40.0:
        level = "Medium"
        recommendation = "Moderate quality concern. Requires manual supervisor review."
    else:
        level = "Low"
        recommendation = "Minor cosmetic defect. Product acceptable for release."

    return {
        "severity_score": overall_score,
        "severity_level": level,
        "recommendation": recommendation,
        "breakdown": {
            "size_score": size_score,
            "location_score": location_score,
            "type_score": type_score,
            "confidence_score": confidence_score
        }
    }

File: frontend/src/utils/severityCalculator.js
/**
 * VisionInspect AI - Severity Scoring Engine
 * Evaluates defects based on the document specification:
 * Severity Score = (Size * 30%) + (Location * 25%) + (Defect Type * 25%) + (Confidence * 20%)
 */

export const calculateSeverityScore = ({
  sizeScore = 0,
  locationScore = 0,
  defectTypeScore = 0,
  confidenceScore = 0,
}) => {
  const sizeContrib = (sizeScore * 30) / 100;
  const locationContrib = (locationScore * 25) / 100;
  const typeContrib = (defectTypeScore * 25) / 100;
  const confidenceContrib = (confidenceScore * 20) / 100;

  const totalScore = Math.round(sizeContrib + locationContrib + typeContrib + confidenceContrib);

  let level = 'Low';
  let badgeColor = 'emerald';
  let recommendation = 'Pass Product (Acceptable Quality)';
  let passVerdict = true;

  if (totalScore >= 80) {
    level = 'Critical';
    badgeColor = 'rose';
    recommendation = 'Reject Product & Trigger Quality Inspection Workflow';
    passVerdict = false;
  } else if (totalScore >= 60) {
    level = 'High';
    badgeColor = 'amber';
    recommendation = 'Repair or Rework Recommended';
    passVerdict = false;
  } else if (totalScore >= 40) {
    level = 'Medium';
    badgeColor = 'yellow';
    recommendation = 'Manual Quality Inspection Review Required';
    passVerdict = false;
  } else {
    level = 'Low';
    badgeColor = 'emerald';
    recommendation = 'Product Passed Quality Check (Minor Cosmetic Defect)';
    passVerdict = true;
  }

  return {
    score: totalScore,
    level,
    badgeColor,
    recommendation,
    passVerdict,
    breakdown: {
      sizeContrib,
      locationContrib,
      typeContrib,
      confidenceContrib,
    },
  };
};

export const DEFECT_TYPE_WEIGHTS = {
  'Surface Scratch': { baseScore: 40, category: 'Cosmetic' },
  'Surface Crack': { baseScore: 85, category: 'Structural' },
  'Missing Component': { baseScore: 95, category: 'Functional' },
  'Pore / Void': { baseScore: 60, category: 'Material' },
  'Solder Bridge / Short': { baseScore: 90, category: 'Electrical' },
  'Discoloration': { baseScore: 25, category: 'Cosmetic' },
};

export const LOCATION_WEIGHTS = {
  'Cosmetic Surface': { score: 30, description: 'Non-critical exterior surface' },
  'Assembly Edge': { score: 60, description: 'Non-functional structural boundary' },
  'Functional Component Area': { score: 90, description: 'High-stress critical operation zone' },
};
